cep-4.md

```diff
@@ -0,0 +1,134 @@
+<table>
+<tr><td> Title </td><td> Implement initial conda plugin mechanism </td>
+<tr><td> Status </td><td> Draft  </td></tr>
+<tr><td> Author(s) </td><td> Bianca Henderson &lt;bhenderson@anaconda.com&gt;</td></tr>
+<tr><td> </td><td>Jannis Leidel &lt;jleidel@anaconda.com&gt; </td></tr>
+<tr><td> Created </td><td> July 5, 2022 </td></tr>
+<tr><td> Updated </td><td> July 6, 2022 </td></tr>
+<tr><td> Discussion </td><td>https://github.com/conda-incubator/ceps/pull/32</td></tr>
+<tr><td> Implementation </td><td>https://github.com/conda/conda/pull/11435</td></tr>
+</table>
+
+## Abstract
+
+In order to enable customization and extra features that are compatible with and discoverable by `conda` (but do not necessarily ship as a default part of the `conda` codebase), we would like to implement an official `conda` plugin mechanism.
+
+A plugin mechanism in `conda` would provide many benefits, including (but not limited to) the coverage of underserved use cases, the added ability to extend `conda` internals via official APIs, as well as lowering the barrier of entry for contributions from other stakeholders in the `conda` community and ecosystem.
+
+The [initial proposal to add this plugin architecture](https://github.com/conda-incubator/ceps/blob/main/cep-2.md) has been [officially approved](https://github.com/conda-incubator/ceps/issues/23); this current CEP will discuss the specific way that the plugin mechanism will actually be implemented.
+
+## Specification
+
+Plugins in `conda` will integrate the "hook + entry point" structure by utilizing the [`pluggy`](https://pluggy.readthedocs.io/en/stable/index.html) Python framework. This implementation can be broken down via the following two steps:
+
+- Define the hook(s) to be registered
+- Register the plugin under the `conda` entrypoint namespace
+
+
+### Hook
+
+Below is an example of a very basic plugin "hook":
+
+_my_plugin.py_
+```python
+import conda.plugins
+
+
+@conda.plugins.register
+def conda_subcommands():
+    ...
+```
+
+### Entry point namespace utilizing a `setup.py` file
+
+Below is an example of an entry point namespace for the custom plugin function, decorated with the plugin hook shown above:
+
+_setup.py_
+```python
+from setuptools import setup
+
+setup(
+    name="my-conda-plugin",
+    install_requires="conda",
+    entry_points={"conda": ["my-conda-plugin = my_plugin"]},
+    py_modules=["my_plugin"],
+)
+```
+
+### Packaging via a `pyproject.toml` file
+
+
+Below is a different packaging example that configures `setuptools` using a `pyproject.toml` file:
+
+_setup.py_
+```python
+import setuptools
+
+if __name__ == "__main__":
+    setuptools.setup()
+```
+
+_pyproject.toml_
+```toml
+[build-system]
+requires = ["setuptools", "setuptools-scm"]
```

Comment:
Based on my personal experience, I would avoid explicitly defining a full example of how to package a project like this and instead focus on how to specify an entry point. Otherwise I could see people copy-and-pasting either example for all their plugins as **the** way to package conda plugins.

I would also list the `pyproject.toml` example first as that's the standard.